{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains all code and steps for the analysis in my BINP37 project. The outline is: \n",
    "- Packages and Set Up\n",
    "- Data Read-in\n",
    "- Quality Control\n",
    "- Doublet Detection\n",
    "- Data Merge\n",
    "- Normalization & Feature Selection\n",
    "- Cell Cycle Analysis\n",
    "- Dimensionality Reducton\n",
    "- Batch Correction methods\n",
    "- Cell Annotation\n",
    "    - Automated\n",
    "    - Manual\n",
    "- Plotting and Visualizations \n",
    "\n",
    "The analysis was informed and shaped by utilizing multiple resources: \n",
    "- Single-cell best practices, https://www.sc-best-practices.org/preamble.html,\n",
    "    - Heumos, L., Schaar, A.C., Lance, C. et al. Best practices for single-cell analysis across modalities. Nat Rev Genet (2023). https://doi.org/10.1038/s41576-023-00586-w\n",
    "- Scanpy's 'Preprocessing and clustering' tutorial, https://scanpy.readthedocs.io/en/stable/tutorials/basics/clustering.html\n",
    "- Orchestrating Single-Cell Analysis with Bioconductor, https://bioconductor.org/books/release/OSCA/book-contents.html\n",
    "- Scverse, community discussions, https://discourse.scverse.org/\n",
    "- NBIS/SciLifeLab Single Cell RNA-seq Analysis Workshop and Tutorials, https://nbisweden.github.io/workshop-scRNAseq/\n",
    "- Biostars Community discussion, https://www.biostars.org/\n",
    "- As well as consulting the documentation for all tools mentioned:\n",
    "    - AnnData\n",
    "    - scanpy\n",
    "    - pandas\n",
    "    - numpy\n",
    "    - matplotlib\n",
    "    - scipy\n",
    "    - Scvi-tools\n",
    "    - Decoupler\n",
    "    - BBKNN\n",
    "    - ComBat\n",
    "    - ACT\n",
    "\n",
    "List of packages and the versions used in this analysis: \n",
    "- scanpy==1.10.1\n",
    "- anndata==0.10.7\n",
    "- umap==0.5.5 \n",
    "- numpy==1.26.4 \n",
    "- scipy==1.12.0 \n",
    "- pandas==2.2.1 \n",
    "- scikit-learn==1.4.2\n",
    "- scikit-misc==0.3.1\n",
    "- scvi-tools==0.14.6\n",
    "- bbknn==1.6.0 \n",
    "- decoupler==1.4.0 \n",
    "\n",
    "\n",
    "\n",
    "The full list of all packages, their dependencies and their versions used:\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Name                    | Version  | Build                   | Channel       |\n",
    "|-------------------------|----------|-------------------------|---------------|\n",
    "| absl-py                 | 2.1.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| anndata                 | 0.10.7   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| annoy                   | 1.17.3   | pypi_0                  | pypi          |\n",
    "| anyio                   | 4.3.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| aom                     | 3.8.2    | h078ce10_0              | conda-forge   |\n",
    "| appnope                 | 0.1.4    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| argon2-cffi             | 23.1.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| argon2-cffi-bindings    | 21.2.0   | py39h0f82c59_4          | conda-forge   |\n",
    "| arpack                  | 3.8.0    | nompi_ha3438d0_101      | conda-forge   |\n",
    "| array-api-compat        | 1.6      | pyhd8ed1ab_0            | conda-forge   |\n",
    "| arrow                   | 1.3.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| asttokens               | 2.4.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| async-lru               | 2.0.4    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| attrs                   | 23.2.0   | pyh71513ae_0            | conda-forge   |\n",
    "| babel                   | 2.14.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| bbknn                   | 1.6.0    | pypi_0                  | pypi          |\n",
    "| beautifulsoup4          | 4.12.3   | pyha770c72_0            | conda-forge   |\n",
    "| bleach                  | 6.1.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| blosc                   | 1.21.5   | h9c252e8_1              | conda-forge   |\n",
    "| brotli                  | 1.1.0    | hb547adb_1              | conda-forge   |\n",
    "| brotli-bin              | 1.1.0    | hb547adb_1              | conda-forge   |\n",
    "| brotli-python           | 1.1.0    | py39hb198ff7_1          | conda-forge   |\n",
    "| brunsli                 | 0.1      | h9f76cd9_0              | conda-forge   |\n",
    "| bzip2                   | 1.0.8    | h93a5062_5              | conda-forge   |\n",
    "| c-ares                  | 1.28.1   | h93a5062_0              | conda-forge   |\n",
    "| c-blosc2                | 2.14.4   | ha57e6be_1              | conda-forge   |\n",
    "| ca-certificates         | 2024.2.2 | hf0a4a13_0              | conda-forge   |\n",
    "| cached-property         | 1.5.2    | hd8ed1ab_1              | conda-forge   |\n",
    "| cached_property         | 1.5.2    | pyha770c72_1            | conda-forge   |\n",
    "| certifi                 | 2024.2.2 | pyhd8ed1ab_0            | conda-forge   |\n",
    "| cffi                    | 1.16.0   | py39he153c15_0          | conda-forge   |\n",
    "| charls                  | 2.4.2    | h13dd4ca_0              | conda-forge   |\n",
    "| charset-normalizer      | 3.3.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| chex                    | 0.1.86   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| click                   | 8.1.7    | unix_pyh707e725_0       | conda-forge   |\n",
    "| cloudpickle             | 3.0.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| colorama                | 0.4.6    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| comm                    | 0.2.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| contourpy               | 1.2.1    | py39h48c5dd5_0          | conda-forge   |\n",
    "| cycler                  | 0.12.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| cython                  | 3.0.10   | pypi_0                  | pypi          |\n",
    "| cytoolz                 | 0.12.3   | py39h17cfd9d_0          | conda-forge   |\n",
    "| dask-core               | 2024.4.2 | pyhd8ed1ab_0            | conda-forge   |\n",
    "| dav1d                   | 1.2.1    | hb547adb_0              | conda-forge   |\n",
    "| debugpy                 | 1.8.1    | py39hf3050f2_0          | conda-forge   |\n",
    "| decorator               | 5.1.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| decoupler               | 1.4.0    | pyhdfd78af_1            | bioconda      |\n",
    "| defusedxml              | 0.7.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| docrep                  | 0.3.2    | pyh44b312d_0            | conda-forge   |\n",
    "| entrypoints             | 0.4      | pyhd8ed1ab_0            | conda-forge   |\n",
    "| et_xmlfile              | 1.1.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| exceptiongroup          | 1.2.0    | pyhd8ed1ab_2            | conda-forge   |\n",
    "| executing               | 2.0.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| fbpca                   | 1.0      | py_0                    | conda-forge   |\n",
    "| filelock                | 3.13.4   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| fonttools               | 4.51.0   | py39h17cfd9d_0          | conda-forge   |\n",
    "| fqdn                    | 1.5.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| freetype                | 2.12.1   | hadb7bae_2              | conda-forge   |\n",
    "| fsspec                  | 2024.3.1 | pyhca7485f_0            | conda-forge   |\n",
    "| future                  | 1.0.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| geosketch               | 1.2      | py_0                    | bioconda      |\n",
    "| get-annotations         | 0.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| giflib                  | 5.2.2    | h93a5062_0              | conda-forge   |\n",
    "| glpk                    | 5.0      | h6d7a090_0              | conda-forge   |\n",
    "| gmp                     | 6.3.0    | hebf3989_1              | conda-forge   |\n",
    "| Name                    | Version  | Build                   | Channel       |\n",
    "|-------------------------|----------|-------------------------|---------------|\n",
    "| gmpy2                   | 2.1.2    | py39h0b4f9c6_1          | conda-forge   |\n",
    "| grpcio                  | 1.60.1   | py39h047a24b_0          | conda-forge   |\n",
    "| h11                     | 0.14.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| h2                      | 4.1.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| h5py                    | 3.10.0   | nompi_py39hd62a535_101  | conda-forge   |\n",
    "| hdf5                    | 1.14.3   | nompi_h5bb55e9_100      | conda-forge   |\n",
    "| hpack                   | 4.0.0    | pyh9f0ad1d_0            | conda-forge   |\n",
    "| httpcore                | 1.0.5    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| httpx                   | 0.27.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| hyperframe              | 6.0.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| icu                     | 73.2     | hc8870d7_0              | conda-forge   |\n",
    "| idna                    | 3.6      | pyhd8ed1ab_0            | conda-forge   |\n",
    "| igraph                  | 0.10.10  | h48be1ad_0              | conda-forge   |\n",
    "| imagecodecs             | 2024.1.1 | py39h2f378ec_6          | conda-forge   |\n",
    "| imageio                 | 2.34.0   | pyh4b66e23_0            | conda-forge   |\n",
    "| importlib-metadata      | 7.1.0    | pyha770c72_0            | conda-forge   |\n",
    "| importlib-resources     | 6.4.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| importlib_metadata      | 7.1.0    | hd8ed1ab_0              | conda-forge   |\n",
    "| importlib_resources     | 6.4.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| inflect                 | 7.2.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| intervaltree            | 3.1.0    | pyhd8ed1ab_1            | conda-forge   |\n",
    "| ipykernel               | 6.29.3   | pyh3cd1d5f_0            | conda-forge   |\n",
    "| ipython                 | 8.18.1   | pyh707e725_3            | conda-forge   |\n",
    "| ipywidgets              | 8.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| isoduration             | 20.11.0  | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jax                     | 0.4.25   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jaxlib                  | 0.4.23   | cpu_py39hcdc4e9d_0      | conda-forge   |\n",
    "| jedi                    | 0.19.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jinja2                  | 3.1.3    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| joblib                  | 1.4.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| json5                   | 0.9.25   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jsonpointer             | 2.4      | py39h2804cbe_3          | conda-forge   |\n",
    "| jsonschema              | 4.21.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jsonschema-specifications | 2023.12.1 | pyhd8ed1ab_0        | conda-forge   |\n",
    "| jsonschema-with-format-nongpl | 4.21.1 | pyhd8ed1ab_0      | conda-forge   |\n",
    "| jupyter                 | 1.0.0    | pyhd8ed1ab_10           | conda-forge   |\n",
    "| jupyter-lsp             | 2.2.5    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyter_client          | 8.6.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyter_console         | 6.6.3    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyter_core            | 5.7.2    | py39h2804cbe_0          | conda-forge   |\n",
    "| jupyter_events          | 0.10.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyter_server          | 2.14.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyter_server_terminals| 0.5.3    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyterlab              | 4.1.6    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyterlab_pygments     | 0.3.0    | pyhd8ed1ab_1            | conda-forge   |\n",
    "| jupyterlab_server       | 2.26.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jupyterlab_widgets      | 3.0.10   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| jxrlib                  | 1.1      | h93a5062_3              | conda-forge   |\n",
    "| kiwisolver              | 1.4.5    | py39hbd775c9_1          | conda-forge   |\n",
    "| krb5                    | 1.21.2   | h92f50d5_0              | conda-forge   |\n",
    "| lazy_loader             | 0.4      | pyhd8ed1ab_0            | conda-forge   |\n",
    "| lcms2                   | 2.16     | ha0e7c42_0              | conda-forge   |\n",
    "| legacy-api-wrap         | 1.4      | pyhd8ed1ab_1            | conda-forge   |\n",
    "| lerc                    | 4.0.0    | h9a09cb3_0              | conda-forge   |\n",
    "| libabseil               | 20230802.1 | cxx17_h13dd4ca_0     | conda-forge   |\n",
    "| libaec                  | 1.1.3    | hebf3989_0              | conda-forge   |\n",
    "| libavif16               | 1.0.4    | hff135a0_2              | conda-forge   |\n",
    "| libblas                 | 3.9.0    | 22_osxarm64_openblas    | conda-forge   |\n",
    "| libbrotlicommon         | 1.1.0    | hb547adb_1              | conda-forge   |\n",
    "| libbrotlidec            | 1.1.0    | hb547adb_1              | conda-forge   |\n",
    "| libbrotlienc            | 1.1.0    | hb547adb_1              | conda-forge   |\n",
    "| libcblas                | 3.9.0    | 22_osxarm64_openblas    | conda-forge   |\n",
    "| libcurl                 | 8.7.1    | h2d989ff_0              | conda-forge   |\n",
    "| libcxx                  | 16.0.6   | h4653b0c_0              | conda-forge   |\n",
    "| libdeflate              | 1.20     | h93a5062_0              | conda-forge   |\n",
    "| libedit                 | 3.1.20191231 | hc8eb9b7_2          | conda-forge   |\n",
    "| libev                   | 4.33     | h93a5062_2              | conda-forge   |\n",
    "| libffi                  | 3.4.2    | h3422bc3_5              | conda-forge   |\n",
    "| libgfortran             | 5.0.0    | 13_2_0_hd922786_3       | conda-forge   |\n",
    "| libgfortran5            | 13.2.0   | hf226fd6_3              | conda-forge   |\n",
    "| libgrpc                 | 1.60.1   | hfc68871_0              | conda-forge   |\n",
    "| libhwloc                | 2.10.0   | default_h52d8fe8_1000   | conda-forge   |\n",
    "| libhwy                  | 1.1.0    | h2ffa867_0              | conda-forge   |\n",
    "| libiconv                | 1.17     | h0d3ecfb_2              | conda-forge   |\n",
    "| libjpeg-turbo           | 3.0.0    | hb547adb_1              | conda-forge   |\n",
    "| libjxl                  | 0.10.1   | h07599a0_1              | conda-forge   |\n",
    "| liblapack               | 3.9.0    | 22_osxarm64_openblas    | conda-forge   |\n",
    "| libllvm14               | 14.0.6   | hd1a9a77_4              | conda-forge   |\n",
    "| libnghttp2              | 1.58.0   | ha4dd798_1              | conda-forge   |\n",
    "| libopenblas             | 0.3.27   | openmp_h6c19121_0       | conda-forge   |\n",
    "| libpng                  | 1.6.43   | h091b4b1_0              | conda-forge   |\n",
    "| libprotobuf             | 4.25.1   | h810fc01_2              | conda-forge   |\n",
    "| libre2-11               | 2023.09.01 | h741fcf5_1           | conda-forge   |\n",
    "| libsodium               | 1.0.18   | h27ca646_1              | conda-forge   |\n",
    "| libsqlite               | 3.45.2   | h091b4b1_0              | conda-forge   |\n",
    "| libssh2                 | 1.11.0   | h7a5bd25_0              | conda-forge   |\n",
    "| libtiff                 | 4.6.0    | h07db509_3              | conda-forge   |\n",
    "| libtorch                | 2.1.2    | cpu_generic_h7a44f25_3  | conda-forge   |\n",
    "| libuv                   | 1.48.0   | h93a5062_0              | conda-forge   |\n",
    "| libwebp-base            | 1.3.2    | h93a5062_1              | conda-forge   |\n",
    "| libxcb                  | 1.15     | hf346824_0              | conda-forge   |\n",
    "| libxml2                 | 2.12.6   | h0d0cfa8_1              | conda-forge   |\n",
    "| libzlib                 | 1.2.13   | h53f4e23_5              | conda-forge   |\n",
    "| libzopfli               | 1.0.3    | h9f76cd9_0              | conda-forge   |\n",
    "| llvm-openmp             | 18.1.3   | hcd81f8e_0              | conda-forge   |\n",
    "| llvmlite                | 0.42.0   | py39h047a24b_1          | conda-forge   |\n",
    "| locket                  | 1.0.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| lz4-c                   | 1.9.4    | hb7217d7_0              | conda-forge   |\n",
    "| markdown                | 3.6      | pyhd8ed1ab_0            | conda-forge   |\n",
    "| markdown-it-py          | 3.0.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| markupsafe              | 2.1.5    | py39h17cfd9d_0          | conda-forge   |\n",
    "| matplotlib-base         | 3.8.4    | py39hbab7938_0          | conda-forge   |\n",
    "| matplotlib-inline       | 0.1.6    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| mdurl                   | 0.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| mistune                 | 3.0.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| ml_dtypes               | 0.4.0    | py39h47e51b9_0          | conda-forge   |\n",
    "| more-itertools          | 10.2.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| mpc                     | 1.3.1    | h91ba8db_0              | conda-forge   |\n",
    "| mpfr                    | 4.2.1    | h41d338b_1              | conda-forge   |\n",
    "| mpmath                  | 1.3.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| munkres                 | 1.1.4    | pyh9f0ad1d_0            | conda-forge   |\n",
    "| natsort                 | 8.4.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| nbclient                | 0.10.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| nbconvert               | 7.16.3   | hd8ed1ab_1              | conda-forge   |\n",
    "| nbconvert-core          | 7.16.3   | pyhd8ed1ab_1            | conda-forge   |\n",
    "| nbconvert-pandoc        | 7.16.3   | hd8ed1ab_1              | conda-forge   |\n",
    "| nbformat                | 5.10.4   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| ncurses                 | 6.4.20240210 | h078ce10_0          | conda-forge   |\n",
    "| nest-asyncio            | 1.6.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| networkx                | 3.2.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| nomkl                   | 1.0      | h5ca1d4c_0              | conda-forge   |\n",
    "| notebook                | 7.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| notebook-shim           | 0.2.4    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| numba                   | 0.59.1   | py39h278f47c_0          | conda-forge   |\n",
    "| numpy                   | 1.26.4   | py39h7aa2656_0          | conda-forge   |\n",
    "| omnipath                | 1.0.8    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| openjpeg                | 2.5.2    | h9f1df11_0              | conda-forge   |\n",
    "| openpyxl                | 3.1.2    | py39h0f82c59_0          | conda-forge   |\n",
    "| openssl                 | 3.3.0    | h0d3ecfb_0              | conda-forge   |\n",
    "| opt-einsum              | 3.3.0    | hd8ed1ab_2              | conda-forge   |\n",
    "| opt_einsum              | 3.3.0    | pyhc1e730c_2            | conda-forge   |\n",
    "| overrides               | 7.7.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| packaging               | 24.0     | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pandas                  | 2.2.1    | py39h47e51b9_0          | conda-forge   |\n",
    "| pandoc                  | 3.1.13   | hce30654_0              | conda-forge   |\n",
    "| pandocfilters           | 1.5.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| parso                   | 0.8.4    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| partd                   | 1.4.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| patsy                   | 0.5.6    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pexpect                 | 4.9.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pickleshare             | 0.7.5    | py_1003                 | conda-forge   |\n",
    "| pillow                  | 10.3.0   | py39h3352c98_0          | conda-forge   |\n",
    "| pip                     | 24.0     | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pkgutil-resolve-name    | 1.3.10   | pyhd8ed1ab_1            | conda-forge   |\n",
    "| platformdirs            | 4.2.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| prometheus_client       | 0.20.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| prompt-toolkit          | 3.0.42   | pyha770c72_0            | conda-forge   |\n",
    "| prompt_toolkit          | 3.0.42   | hd8ed1ab_0              | conda-forge   |\n",
    "| protobuf                | 4.25.1   | py39haf112ec_0          | conda-forge   |\n",
    "| psutil                  | 5.9.8    | py39h17cfd9d_0          | conda-forge   |\n",
    "| pthread-stubs           | 0.4      | h27ca646_1001           | conda-forge   |\n",
    "| ptyprocess              | 0.7.0    | pyhd3deb0d_0            | conda-forge   |\n",
    "| pure_eval               | 0.2.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pycparser               | 2.22     | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pydeprecate             | 0.3.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pygments                | 2.17.2   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pynndescent             | 0.5.12   | pyhca7485f_0            | conda-forge   |\n",
    "| pyobjc-core             | 10.2     | py39hb167abd_0          | conda-forge   |\n",
    "| pyobjc-framework-cocoa  | 10.2     | py39hb167abd_0          | conda-forge   |\n",
    "| pyparsing               | 3.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pyro-api                | 0.1.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pyro-ppl                | 1.9.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pysocks                 | 1.7.1    | pyha2e5f31_6            | conda-forge   |\n",
    "| python                  | 3.9.19   | hd7ebdb9_0_cpython      | conda-forge   |\n",
    "| python-annoy            | 1.17.2   | py39hb198ff7_1          | conda-forge   |\n",
    "| python-dateutil         | 2.9.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| python-fastjsonschema   | 2.19.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| python-igraph           | 0.11.4   | py39h4c42df0_0          | conda-forge   |\n",
    "| python-json-logger      | 2.0.7    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| python-tzdata           | 2024.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| python_abi              | 3.9      | 4_cp39                  | conda-forge   |\n",
    "| pytorch                 | 2.1.2    | cpu_generic_py39h4bad755_3 | conda-forge |\n",
    "| pytorch-lightning       | 1.3.8    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pytz                    | 2024.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| pywavelets              | 1.4.1    | py39hf4a74a7_1          | conda-forge   |\n",
    "| pyyaml                  | 5.4.1    | py39h02fc5c5_4          | conda-forge   |\n",
    "| pyzmq                   | 25.1.2   | py39he1e2164_0          | conda-forge   |\n",
    "| qtconsole-base          | 5.5.1    | pyha770c72_0            | conda-forge   |\n",
    "| qtpy                    | 2.4.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| rav1e                   | 0.6.6    | h69fbcac_2              | conda-forge   |\n",
    "| re2                     | 2023.09.01 | h4cba328_1            | conda-forge   |\n",
    "| readline                | 8.2      | h92ec313_1              | conda-forge   |\n",
    "| referencing             | 0.34.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| requests                | 2.31.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| rfc3339-validator       | 0.1.4    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| rfc3986-validator       | 0.1.1    | pyh9f0ad1d_0            | conda-forge   |\n",
    "| rich                    | 13.7.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| rpds-py                 | 0.18.0   | py39h9a407ce_0          | conda-forge   |\n",
    "| scanorama               | 1.7.4    | pyhdfd78af_0            | bioconda      |\n",
    "| scanpy                  | 1.10.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| scikit-image            | 0.19.3   | py39hde7b980_2          | conda-forge   |\n",
    "| scikit-learn            | 1.4.2    | py39h6dd658b_0          | conda-forge   |\n",
    "| scikit-misc             | 0.3.1    | pypi_0                  | pypi          |\n",
    "| scipy                   | 1.12.0   | py39hcc04109_2          | conda-forge   |\n",
    "| scvi-tools              | 0.14.6   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| seaborn                 | 0.13.2   | hd8ed1ab_2              | conda-forge   |\n",
    "| seaborn-base            | 0.13.2   | pyhd8ed1ab_2            | conda-forge   |\n",
    "| send2trash              | 1.8.3    | pyh31c8845_0            | conda-forge   |\n",
    "| session-info            | 1.0.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| setuptools              | 59.5.0   | py39h2804cbe_0          | conda-forge   |\n",
    "| six                     | 1.16.0   | pyh6c4a22f_0            | conda-forge   |\n",
    "| sleef                   | 3.5.1    | h156473d_2              | conda-forge   |\n",
    "| snappy                  | 1.2.0    | hd04f947_1              | conda-forge   |\n",
    "| sniffio                 | 1.3.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| sortedcontainers        | 2.4.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| soupsieve               | 2.5      | pyhd8ed1ab_1            | conda-forge   |\n",
    "| stack_data              | 0.6.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| statsmodels             | 0.14.1   | py39h373d45f_0          | conda-forge   |\n",
    "| stdlib-list             | 0.10.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| svt-av1                 | 2.0.0    | h078ce10_0              | conda-forge   |\n",
    "| sympy                   | 1.12     | pypyh9d50eac_103        | conda-forge   |\n",
    "| tbb                     | 2021.12.0 | h2ffa867_0            | conda-forge   |\n",
    "| tensorboard             | 2.16.2   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| tensorboard-data-server | 0.7.0    | py39had97604_1          | conda-forge   |\n",
    "| terminado               | 0.18.1   | pyh31c8845_0            | conda-forge   |\n",
    "| texttable               | 1.7.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| threadpoolctl           | 3.4.0    | pyhc1e730c_0            | conda-forge   |\n",
    "| tifffile                | 2024.2.12 | pyhd8ed1ab_0          | conda-forge   |\n",
    "| tinycss2                | 1.2.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| tk                      | 8.6.13   | h5083fa2_1              | conda-forge   |\n",
    "| tomli                   | 2.0.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| toolz                   | 0.12.1   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| torchmetrics            | 0.7.3    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| tornado                 | 6.4      | py39h17cfd9d_0          | conda-forge   |\n",
    "| tqdm                    | 4.66.2   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| traitlets               | 5.14.2   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| typeguard               | 4.2.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| types-python-dateutil   | 2.9.0.20240316 | pyhd8ed1ab_0     | conda-forge   |\n",
    "| typing-extensions       | 4.11.0   | hd8ed1ab_0              | conda-forge   |\n",
    "| typing_extensions       | 4.11.0   | pyha770c72_0            | conda-forge   |\n",
    "| typing_utils            | 0.1.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| tzdata                  | 2024a    | h0c530f3_0              | conda-forge   |\n",
    "| umap-learn              | 0.5.5    | py39h2804cbe_1          | conda-forge   |\n",
    "| unicodedata2            | 15.1.0   | py39h0f82c59_0          | conda-forge   |\n",
    "| uri-template            | 1.3.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| urllib3                 | 2.2.1    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| wcwidth                 | 0.2.13   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| webcolors               | 1.13     | pyhd8ed1ab_0            | conda-forge   |\n",
    "| webencodings            | 0.5.1    | pyhd8ed1ab_2            | conda-forge   |\n",
    "| websocket-client        | 1.7.0    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| werkzeug                | 3.0.2    | pyhd8ed1ab_0            | conda-forge   |\n",
    "| wheel                   | 0.43.0   | pyhd8ed1ab_1            | conda-forge   |\n",
    "| widgetsnbextension      | 4.0.10   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| wrapt                   | 1.16.0   | py39h17cfd9d_0          | conda-forge   |\n",
    "| xorg-libxau             | 1.0.11   | hb547adb_0              | conda-forge   |\n",
    "| xorg-libxdmcp           | 1.1.3    | h27ca646_0              | conda-forge   |\n",
    "| xz                      | 5.2.6    | h57fd34a_0              | conda-forge   |\n",
    "| yaml                    | 0.2.5    | h3422bc3_2              | conda-forge   |\n",
    "| zeromq                  | 4.3.5    | hebf3989_1              | conda-forge   |\n",
    "| zfp                     | 1.0.1    | ha8f4885_0              | conda-forge   |\n",
    "| zipp                    | 3.17.0   | pyhd8ed1ab_0            | conda-forge   |\n",
    "| zlib-ng                 | 2.0.7    | h1a8c8d9_0              | conda-forge   |\n",
    "| zstd                    | 1.5.5    | h4f39d0f_0              | conda-forge   |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Packages and Set Up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gc\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scvi\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy\n",
    "import decoupler as dc\n",
    "from bbknn import bbknn\n",
    "from scipy.stats import median_abs_deviation\n",
    "from PIL import Image\n",
    "import matplotlib.gridspec as gridspec"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "dc.show_resources()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3 \n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor=\"white\") \n",
    "results_file = \"write/final_adata.h5ad\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Read in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read each sequencing run data into a dictionary, with a name for each as the key\n",
    "adatas = {\n",
    "    \"D1\" : sc.read_10x_mtx(\n",
    "        \"singlecell_processed/result/Isol_Microglia_EFAD_TD_august/outs/filtered_feature_bc_matrix\",\n",
    "        var_names=\"gene_symbols\",\n",
    "        cache=False,\n",
    "        make_unique=True),\n",
    "    \"D2\" : sc.read_10x_mtx(\n",
    "        \"singlecell_processed/result/collect_tube_1_batch_3_June/outs/filtered_feature_bc_matrix\",\n",
    "        var_names=\"gene_symbols\",\n",
    "        cache=False,\n",
    "        make_unique=True),\n",
    "    \"D3\" : sc.read_10x_mtx(\n",
    "        \"singlecell_processed/result/Tube2_July_20_TD_YYRFC/outs/filtered_feature_bc_matrix\",\n",
    "        var_names=\"gene_symbols\",\n",
    "        cache=False,\n",
    "        make_unique=True),\n",
    "    \"D4\" : sc.read_10x_mtx(\n",
    "        \"singlecell_processed/result/Tube3_July_20_TD_YYRFC/outs/filtered_feature_bc_matrix\",\n",
    "        var_names=\"gene_symbols\",\n",
    "        cache=False,\n",
    "        make_unique=True)\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function Definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function which prints basic Quality Control metrics without any filtering/changing data\n",
    "def summarize_adata(adata):\n",
    "    # Calculate the total number of cells and genes\n",
    "    total_cells = adata.n_obs\n",
    "    total_genes = adata.n_vars\n",
    "    \n",
    "    # Calculate number of cells with fewer than 500 UMI counts\n",
    "    cells_with_500_or_more = (adata.X.sum(axis=1) >= 500).sum()\n",
    "    cells_with_less_than_500 = total_cells - cells_with_500_or_more\n",
    "    print(f\"Number of cells with fewer than 500 UMI counts: {cells_with_less_than_500}\")\n",
    "\n",
    "    # Calculate number of cells with fewer than 200 genes\n",
    "    cells_with_200_or_more_genes = (adata.X > 0).sum(axis=1) >= 200\n",
    "    cells_with_less_than_200_genes = total_cells - cells_with_200_or_more_genes.sum()\n",
    "    print(f\"Number of cells with fewer than 200 genes: {cells_with_less_than_200_genes}\")\n",
    "\n",
    "    # Calculate number of genes expressed in fewer than 3 cells\n",
    "    genes_with_3_or_more_cells = (adata.X > 0).sum(axis=0) >= 3\n",
    "    genes_with_less_than_3_cells = total_genes - genes_with_3_or_more_cells.sum()\n",
    "    print(f\"Number of genes expressed in fewer than 3 cells: {genes_with_less_than_3_cells}\")\n",
    "    \n",
    "\n",
    "    # Print current status of the AnnData object\n",
    "    print(f\"Current Anndata has {adata.n_obs} cells and {adata.n_vars} genes, with a total amount of {adata.X.sum()} UMI counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def QC_filter_adata(adata):\n",
    "    # Annotate the group of mitochondrial genes as 'mt'\n",
    "    adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "\n",
    "    # Calculate QC metrics\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "    # Visualize QC metrics before filtering\n",
    "    print(\"Visualizations before filtering:\")\n",
    "    sc.pl.violin(adata, [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"], jitter=0.4, multi_panel=True)\n",
    "    sc.pl.scatter(adata, x=\"total_counts\", y=\"n_genes_by_counts\", color=\"pct_counts_mt\")\n",
    "\n",
    "    # Print the number of cells with more than 5% mitochondrial genes\n",
    "    cells_with_more_than_5_percent_mt = (adata.obs[\"pct_counts_mt\"] > 5).sum()\n",
    "    print(f\"{cells_with_more_than_5_percent_mt} of {adata.n_obs} cells contain more than 5% mitochondrial genes. Will be filtered\")\n",
    "\n",
    "    # Filter cells with more than 5% mitochondrial genes\n",
    "    adata = adata[adata.obs.pct_counts_mt < 5, :]\n",
    "    sc.pp.filter_cells(adata, min_counts=500)\n",
    "    sc.pp.filter_cells(adata, min_genes=200)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "    # Recalculate QC metrics for the filtered data\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "    # Visualize QC metrics after filtering\n",
    "    print(\"Visualizations after filtering:\")\n",
    "    sc.pl.violin(adata, [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"], jitter=0.4, multi_panel=True)\n",
    "    sc.pl.scatter(adata, x=\"total_counts\", y=\"n_genes_by_counts\", color=\"pct_counts_mt\")\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_cellcycle_mouse(adata):\n",
    "    # Get cell cycle genes from the regev lab data file. sort/split into appropriate lists\n",
    "    cell_cycle_genes = [x.strip().lower().title() for x in open('./regev_lab_cell_cycle_genes.txt')]\n",
    "    s_genes = cell_cycle_genes[:43]\n",
    "    g2m_genes = cell_cycle_genes[43:]\n",
    "    cell_cycle_genes = [x.lower().title() for x in cell_cycle_genes if x.lower().title() in adata.var_names]\n",
    "    # Cell cycle scoring function from scanpy\n",
    "    sc.tl.score_genes_cell_cycle(adata, s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "\n",
    "    # subset adata with only necessary genes\n",
    "    adata_cc_genes = adata[:, cell_cycle_genes]\n",
    "    \n",
    "\n",
    "    adata.obs[\"cellcycle\"] = adata_cc_genes.obs[\"phase\"]\n",
    "    adata.obs[\"S_score\"] = adata_cc_genes.obs[\"S_score\"]\n",
    "    adata.obs[\"G2M_score\"] = adata_cc_genes.obs[\"G2M_score\"]\n",
    "    \n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to generate a dictionary of lists of genes relevant to analysis which are present in the dataset\n",
    "def find_genes(adata):\n",
    "    #define some lists of genes of interest for the analysis.\n",
    "    Genes_of_interest = []\n",
    "    homeo_genes = ['Sall1', 'Cx3cr1', 'P2ry12', 'P2ry13', 'Olfml3', 'Tmem119', 'Cd68', 'Itgam', 'Cst3']\n",
    "    DAM_genes = ['Apoe', 'B2m', 'Ctsb', 'Lpl', 'Cst7', 'Tyrobp', 'Trem2', 'Cd9', 'Itgax', 'Cd63', 'Fth1', 'Spp1', 'Axl', 'Mertk', 'Mdk', 'Ptn']\n",
    "    Mapk_genes = ['Syk', 'Prkca', 'Mef2c', 'Elk4', 'Trp53', 'Mertk', 'Axl', 'Mapk14', 'Mapk1', 'Gadd45a',]\n",
    "    Nfkb_genes = ['Nfkb1', 'Tlr4', 'Nfkbia', 'Cd40', 'Tlr4', 'Rela', 'Relb']\n",
    "    Macroph_genes = ['Mrc1', 'Mrc2', 'Ccr2', 'Ly6c2', 'Lyz2', 'Vim', 'Ifi204', 'S100a10', 'Msrb1']\n",
    "    # make a dictionary of the lists to iterate through\n",
    "    gene_lists = {\"Homeo\": homeo_genes, \"DAM\": DAM_genes, \"Mapk\": Mapk_genes, \"Nfkb\": Nfkb_genes, \"Macroph\": Macroph_genes}\n",
    "    found_genes = {key: [] for key in gene_lists.keys()}\n",
    "\n",
    "    # loop through gene lists and genes, see if it is present in the passed adata, if so append it to a list+dictionary \n",
    "    # so we have a dicitonary of lists of only genes in the dataset, as well as a list of all genes from the different types\n",
    "    for gene_list_name, gene_list in gene_lists.items():\n",
    "        for gene in gene_list:\n",
    "            if gene in adata.var.index:\n",
    "                print(f\" ✓ {gene} is in adata.var\")\n",
    "                Genes_of_interest.append(gene)\n",
    "                found_genes[gene_list_name].append(gene)\n",
    "            else:\n",
    "                print(f\"{gene} is not in adata.var\")\n",
    "\n",
    "    return Genes_of_interest, found_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to perform leiden clustering on a dataset through a specified range of resolutions\n",
    "# Didnt end up being as useful as I wouldve liked due to Anndata+dicitonary behaviour\n",
    "def leiden_clustering_range(adata, start: float = 0.01, end: float = 2.1, step: float = 0.01):\n",
    "\n",
    "    # Initialize an empty dictionary to store clustering results\n",
    "    leiden_clustering_results = {}\n",
    "\n",
    "    # Iterate over the desired range of resolutions\n",
    "    i = start\n",
    "    while i <= end:\n",
    "        sc.tl.leiden(\n",
    "            adata,\n",
    "            resolution=i,\n",
    "            random_state=0,\n",
    "            n_iterations=2,\n",
    "            directed=False,\n",
    "            key_added=f\"leiden_{i}\",\n",
    "            flavor=\"igraph\"\n",
    "        )\n",
    "        \n",
    "        # Store the clustering results in the dictionary\n",
    "        leiden_clustering_results[round(i, 2)] = adata.obs[f\"leiden_{i}\"].copy()\n",
    "        \n",
    "        # Remove the individual leiden_{i} from .obs and .uns\n",
    "        if f\"leiden_{i}\" in adata.obs:\n",
    "            del adata.obs[f\"leiden_{i}\"]\n",
    "        if f\"leiden_{i}\" in adata.uns:\n",
    "            del adata.uns[f\"leiden_{i}\"]\n",
    "\n",
    "        # Increment the resolution\n",
    "        i += step\n",
    "        i = round(i, 2)\n",
    "\n",
    "    # Store the dictionary in both the 'uns' and 'obs' attributes of adata\n",
    "    adata.uns[\"leiden_clustering_results\"] = leiden_clustering_results\n",
    "    adata.obs[\"leiden_clustering_results\"] = leiden_clustering_results\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to plot UMAPs of leiden clustering for specified resolutions\n",
    "def plot_leiden(adata, resolutions):\n",
    "    for resolution in resolutions:\n",
    "        # Extract the cluster labels for the given resolution\n",
    "        leiden_clusters = adata.uns[\"leiden_clustering_results\"][resolution]\n",
    "\n",
    "        # Temporarily add the cluster labels to obs\n",
    "        temp_column_name = f\"leiden_{resolution}\"\n",
    "        adata.obs[temp_column_name] = leiden_clusters\n",
    "\n",
    "        # Plot the UMAP with the temporary column\n",
    "        sc.pl.umap(adata, color=temp_column_name, legend_loc=\"on data\", title=f\"Leiden Clustering (resolution={resolution})\")\n",
    "\n",
    "        # Remove the temporary column\n",
    "        del adata.obs[temp_column_name]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print Quality Control metric for each dataset\n",
    "for adata_key, adata in adatas.items():\n",
    "    print(f\"{adata_key}:\")\n",
    "    summarize_adata(adata)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_adatas = {}\n",
    "QC_adatas = {}\n",
    "for adata_key, adata in adatas.items():\n",
    "    print(f\"{adata_key}:\")\n",
    "    filtered_adatas[adata_key] = QC_filter_adata(adata)\n",
    "    QC_adatas[adata_key] = QC_filter_adata(adata)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(filtered_adatas)\n",
    "print(QC_adatas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Doublet Detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for adata_key, adata in filtered_adatas.items():\n",
    "    sc.pp.highly_variable_genes(adata, n_top_genes = 3000, subset = True, flavor = 'seurat_v3')\n",
    "    print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae_models = {}\n",
    "solo_models = {}\n",
    "\n",
    "# Define directories to save the models\n",
    "vae_model_dir = \"vae_models\"\n",
    "solo_model_dir = \"solo_models\"\n",
    "\n",
    "# Create directories if they don't exist\n",
    "os.makedirs(vae_model_dir, exist_ok=True)\n",
    "os.makedirs(solo_model_dir, exist_ok=True)\n",
    "\n",
    "# Load or train models\n",
    "for adata_key, adata in filtered_adatas.items():\n",
    "    vae_model_path = os.path.join(vae_model_dir, f\"{adata_key}_vae\")\n",
    "    solo_model_path = os.path.join(solo_model_dir, f\"{adata_key}_solo\")\n",
    "    \n",
    "    if os.path.exists(vae_model_path):\n",
    "        # Load VAE model if it exists\n",
    "        vae_models[adata_key] = scvi.model.SCVI.load(vae_model_path, adata)\n",
    "    else:\n",
    "        # Setup and train VAE model if it doesn't exist\n",
    "        scvi.model.SCVI.setup_anndata(adata)\n",
    "        vae_models[adata_key] = scvi.model.SCVI(adata)\n",
    "        vae_models[adata_key].train()\n",
    "        vae_models[adata_key].save(vae_model_path)\n",
    "\n",
    "    if os.path.exists(solo_model_path):\n",
    "        # Load SOLO model if it exists\n",
    "        solo_models[adata_key] = scvi.external.SOLO.load(solo_model_path, adata)\n",
    "    else:\n",
    "        # Setup and train SOLO model if it doesn't exist\n",
    "        solo_models[adata_key] = scvi.external.SOLO.from_scvi_model(vae_models[adata_key])\n",
    "        solo_models[adata_key].train()\n",
    "        solo_models[adata_key].save(solo_model_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_df = {}\n",
    "for batch, solo in solo_models.items():\n",
    "    batch_df[batch] = solo.predict()\n",
    "    batch_df[batch]['solo_prediction'] = solo.predict(soft = False)\n",
    "    batch_df[batch].index = batch_df[batch].index.map(lambda x: x[:-2])\n",
    "    if batch in QC_adatas:\n",
    "        QC_adatas[batch].obs['solo_prediction'] = batch_df[batch]['solo_prediction']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for batch in batch_df:\n",
    "    doublets = len(batch_df[batch][batch_df[batch].solo_prediction == 'doublet'])\n",
    "    total = len(batch_df[batch])\n",
    "    percentage = (doublets / total) * 100\n",
    "    print(f\"The number of doublets detected is: {doublets}, which is {percentage:.2f}% of cells.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "le = LabelEncoder()\n",
    "n = len(batch_df)\n",
    "plt.figure(figsize=(n*5, 5)) \n",
    "\n",
    "for i, batch in enumerate(batch_df, start=1): \n",
    "    colors = le.fit_transform(batch_df[batch]['solo_prediction'])\n",
    "    plt.subplot(1, n, i)\n",
    "    plt.scatter(batch_df[batch]['doublet'], batch_df[batch]['singlet'], c=colors, cmap='viridis', s=.5)\n",
    "    plt.xlabel('Doublet')\n",
    "    plt.ylabel('Singlet')\n",
    "    plt.title(f'Batch: {batch}')\n",
    "    plt.xlim(-6, 3)  # Set the limits of the x-axis\n",
    "    plt.ylim(-3, 6)  # Set the limits of the y-axis\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for adata_key, adata in QC_adatas.items():\n",
    "    sc.pp.scrublet(adata)\n",
    "    adata.obs.rename(columns={'predicted_doublet': 'scrublet_prediction'}, inplace=True)\n",
    "    print(f\"\\n Scrublet found {adata.obs['scrublet_prediction'].sum()} doublets out of {adata.n_obs} cells \\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Merge "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label, data in QC_adatas.items():\n",
    "    # Add the label as a prefix to the observation names\n",
    "    data.obs_names = label + \"_\" + data.obs_names.astype(str)\n",
    "    print(label)\n",
    "\n",
    "adata = ad.concat(QC_adatas, label=\"sample\", join=\"outer\", merge=\"unique\")\n",
    "\n",
    "genotype_dict = {\"D4_\": \"FAD\", \"D2_\": \"FAD\", \"D3_\": \"5/8_WT\", \"D1_\": \"5/8_WT\"}\n",
    "treatment_dict = {\"D4_\": \"LPS\", \"D2_\": \"Vehicle\", \"D3_\": \"LPS\", \"D1_\": \"Vehicle\"}\n",
    "\n",
    "adata.obs[\"genotype\"] = adata.obs[\"sample\"].map(lambda x: genotype_dict[x.split(\"_\")[0] + \"_\"])\n",
    "adata.obs[\"treatment\"] = adata.obs[\"sample\"].map(lambda x: treatment_dict[x.split(\"_\")[0] + \"_\"])\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for label, data in QC_adatas.items():\n",
    "    # Add the label as a prefix to the observation names\n",
    "    data.obs_names = label + \"_\" + data.obs_names.astype(str)\n",
    "\n",
    "scvi_adata = ad.concat(QC_adatas, label=\"sample\", axis=0, join=\"inner\", merge=\"unique\")\n",
    "\n",
    "\n",
    "genotype_dict = {\"D4_\": \"FAD\", \"D2_\": \"FAD\", \"D3_\": \"5/8_WT\", \"D1_\": \"5/8_WT\"}\n",
    "treatment_dict = {\"D4_\": \"LPS\", \"D2_\": \"Vehicle\", \"D3_\": \"LPS\", \"D1_\": \"Vehicle\"}\n",
    "\n",
    "scvi_adata.obs[\"genotype\"] = scvi_adata.obs[\"sample\"].map(lambda x: genotype_dict[x.split(\"_\")[0] + \"_\"])\n",
    "scvi_adata.obs[\"treatment\"] = scvi_adata.obs[\"sample\"].map(lambda x: treatment_dict[x.split(\"_\")[0] + \"_\"])\n",
    "\n",
    "scvi_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization & Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(adata.obs[\"total_counts\"], bins=100, log=False, label=\"Total counts in bins of 100\")\n",
    "plt.xlabel(\"Total counts\")\n",
    "plt.ylabel(\"Number of cells\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we store the raw (un-normalized) counts in a different layer called 'counts'\n",
    "adata.layers[\"counts\"] = adata.X.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalize and log the data and store that in a different layer too incase we need to switch between them.\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "adata.layers['log_norm'] = adata.X.copy()\n",
    "#so we can plot a histogram of the total counts/cells after norm\n",
    "adata.obs[\"norm_total_counts\"] = adata.X.sum(axis=1)\n",
    "plt.hist(adata.obs[\"norm_total_counts\"], bins=100, log=False, label=\"normed total counts in bins of 100\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We get the highly variable genes using the parameter batch_key which is described in the documentation as follows: [If specified, highly-variable genes are selected within each batch separately and merged. This simple process avoids the selection of batch-specific genes and acts as a lightweight batch correction method](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.highly_variable_genes.html).\n",
    "We plot the HVGs, the scanpy function expects normalized/logarithmized data and by default plots before/after. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata, batch_key=\"sample\")\n",
    "sc.pl.highly_variable_genes(adata)\n",
    "#make some variables to print a nice little summary\n",
    "num_highly_variable_genes_t = adata.var['highly_variable'].sum()\n",
    "total_genes_t = adata.var.shape[0]\n",
    "\n",
    "print(f\"{num_highly_variable_genes_t} out of {total_genes_t} genes are considered highly variable.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Genes_of_interest, marker_genes = find_genes(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#getting ribosomal genes\n",
    "# Create a mask for var_names that start with \"Rpl\" or \"Rps\"\n",
    "mask = adata.var_names.str.startswith('Rpl') | adata.var_names.str.startswith('Rps') | adata.var_names.str.startswith('Mrps') | adata.var_names.str.startswith('Mrpl')\n",
    "\n",
    "# Get the var_names that start with \"Rpl\" or \"Rps\" and convert them to a list\n",
    "ribo_genes = adata.var_names[mask].tolist()\n",
    "print(ribo_genes)\n",
    "print(len(ribo_genes))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Cycle Analysis\n",
    "\n",
    "Next we use the cell cycle scoring method from scanpy using markers from regev lab. and print porportion and porportion by sample "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers[\"counts\"].copy()\n",
    "adata=annotate_cellcycle_mouse(adata)\n",
    "\n",
    "for p in adata.obs['phase'].unique():\n",
    "        phase_count = len(adata.obs[adata.obs['phase'] == p])\n",
    "        total_count = adata.obs.shape[0]\n",
    "        print(f\"The percent of {p} phase in adata is {round(phase_count / total_count * 100, 2)}%\")\n",
    "print(\"\\n\")\n",
    "\n",
    "for s in adata.obs['sample'].unique():\n",
    "    for p in adata.obs['phase'].unique():\n",
    "        phase_count = adata[(adata.obs['phase'] == p) & (adata.obs['sample'] == s)].shape[0]\n",
    "        total_count = adata[adata.obs['sample'] == s].shape[0]\n",
    "        print(f\"The percent of {p} phase in {s} is {round(phase_count / total_count * 100, 2)}%\")\n",
    "    print('\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dimensionality Reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['log_norm'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.pca(adata)\n",
    "sc.pl.pca_variance_ratio(adata, n_pcs=50, log=True)\n",
    "\n",
    "sc.pl.pca(\n",
    "    adata,\n",
    "    color=[\"sample\", \"sample\", \"pct_counts_mt\", \"pct_counts_mt\", \"genotype\", \"genotype\", \"treatment\", \"treatment\",],\n",
    "    dimensions=[(0, 1), (2, 3), (0, 1), (2, 3), (0, 1), (2, 3), (0, 1), (2, 3)],\n",
    "    ncols=2,\n",
    "    size=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_pcs=15)\n",
    "\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pl.umap(adata, color=[\"sample\", \"genotype\"])\n",
    "\n",
    "sc.pl.umap(adata, color=\"pct_counts_mt\")\n",
    "\n",
    "sc.pl.umap(adata, color=\"n_genes_by_counts\")\n",
    "\n",
    "sc.pl.umap(adata, color=\"total_counts\")\n",
    "\n",
    "#sc.pl.umap(adata, color=\"solo_prediction\")\n",
    "\n",
    "sc.pl.umap(adata, color=[\"scrublet_prediction\", \"doublet_score\"])\n",
    "\n",
    "sc.pl.umap(adata, color=marker_genes['Homeo'], vmax=4)\n",
    "\n",
    "sc.pl.umap(adata, color=marker_genes['Macroph'], vmax=4)\n",
    "\n",
    "sc.pl.umap(adata, color=marker_genes['DAM'], vmax=4)\n",
    "\n",
    "sc.pl.umap(adata, color= [\"genotype\", \"treatment\"], wspace= 0.5, legend_loc=\"on data\")\n",
    "\n",
    "sc.pl.umap(adata, color=marker_genes['Mapk'], vmax=4)\n",
    "\n",
    "sc.pl.umap(adata, color=marker_genes['Nfkb'], vmax=4)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch Correction \n",
    "### SCVI batch correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocessing steps\n",
    "scvi_adata.layers[\"counts\"] = scvi_adata.X.copy()  # preserve counts\n",
    "sc.pp.normalize_total(scvi_adata, target_sum=1e4)\n",
    "sc.pp.log1p(scvi_adata)\n",
    "scvi_adata.raw = scvi_adata\n",
    "sc.pp.filter_cells(scvi_adata, min_genes=200)\n",
    "sc.pp.filter_genes(scvi_adata, min_cells=2)\n",
    "scvi_adata.var[\"mt\"] = scvi_adata.var_names.str.startswith(\"mt-\")  # annotate the group of mitochondrial genes as 'mt', mouse mitochondrial genes start with mt\n",
    "sc.pp.calculate_qc_metrics(scvi_adata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True)\n",
    "sc.pp.highly_variable_genes(scvi_adata, batch_key=\"sample\")\n",
    "\n",
    "# Setup anndata for SCVI\n",
    "scvi.model.SCVI.setup_anndata(\n",
    "    scvi_adata,\n",
    "    layer=\"counts\",\n",
    "    batch_key=\"sample\",\n",
    ")\n",
    "\n",
    "# Initialize the model\n",
    "model = scvi.model.SCVI(scvi_adata)\n",
    "\n",
    "# Check if the model is already trained\n",
    "model_path = \"scvi_merged_model_f\"\n",
    "if not os.path.exists(model_path):\n",
    "    # Train the model if it's not trained yet\n",
    "    model.train(check_val_every_n_epoch=1, max_epochs=300, early_stopping=True, early_stopping_patience=10)\n",
    "    # Save the model\n",
    "    model.save(model_path, save_anndata=True)\n",
    "else:\n",
    "    # Load the model if it's already trained\n",
    "    model = scvi.model.SCVI.load(model_path, scvi_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_elbo = model.history[\"elbo_train\"][1:]\n",
    "test_elbo = model.history[\"elbo_validation\"]\n",
    "ax = train_elbo.plot()\n",
    "test_elbo.plot(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi_adata.obsm[\"X_scvi\"] = model.get_latent_representation()\n",
    "scvi_adata.layers[\"scvi_expr\"] = model.get_normalized_expression(scvi_adata, return_mean=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(5, 1, figsize=(5, 25))\n",
    "\n",
    "# Compute neighbors with different parameters\n",
    "sc.pp.neighbors(scvi_adata, use_rep=\"X_scvi\", metric=\"correlation\")\n",
    "sc.tl.umap(scvi_adata)\n",
    "sc.pl.umap(scvi_adata, color=\"sample\", ax=axs[0], show=False)\n",
    "\n",
    "sc.pp.neighbors(scvi_adata, use_rep=\"X_scvi\", metric=\"euclidean\")\n",
    "sc.tl.umap(scvi_adata)\n",
    "sc.pl.umap(scvi_adata, color=\"sample\", ax=axs[1], show=False)\n",
    "\n",
    "sc.pp.neighbors(scvi_adata, use_rep=\"X_scvi\")\n",
    "sc.tl.umap(scvi_adata)\n",
    "sc.pl.umap(scvi_adata, color=\"sample\", ax=axs[2], show=False)\n",
    "\n",
    "sc.pp.neighbors(scvi_adata, use_rep=\"X_scvi\", n_pcs=8)\n",
    "sc.tl.umap(scvi_adata)\n",
    "sc.pl.umap(scvi_adata, color=\"sample\", ax=axs[3], show=False)\n",
    "\n",
    "sc.pp.neighbors(scvi_adata, use_rep=\"X_scvi\", n_pcs=10)\n",
    "sc.tl.umap(scvi_adata)\n",
    "sc.pl.umap(scvi_adata, color=\"sample\", ax=axs[4], show=False)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BBKNN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bbknn_adata = adata.copy()\n",
    "bbknn_adata.X = bbknn_adata.layers['log_norm'].copy()\n",
    "sc.pp.pca(bbknn_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bbknn(bbknn_adata,batch_key='sample', n_pcs=15)\n",
    "sc.tl.umap(bbknn_adata)\n",
    "sc.pl.umap(bbknn_adata,color='sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a new object with lognormalized counts\n",
    "adata_combat = adata.copy() \n",
    "\n",
    "# first store the raw data \n",
    "adata_combat.raw = adata_combat\n",
    "\n",
    "# run combat\n",
    "sc.pp.combat(adata_combat, key='sample')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pp.highly_variable_genes(adata_combat)\n",
    "print(\"Highly variable genes: %d\"%sum(adata_combat.var.highly_variable))\n",
    "sc.pl.highly_variable_genes(adata_combat)\n",
    "\n",
    "sc.pp.pca(adata_combat, mask_var=\"highly_variable\", svd_solver='arpack')\n",
    "\n",
    "sc.pp.neighbors(adata_combat, n_pcs=15)\n",
    "\n",
    "sc.tl.umap(adata_combat)\n",
    "sc.pl.umap(adata_combat, color='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combat_Genes_of_interest, combat_marker_genes = find_genes(adata_combat)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_combat, color=\"total_counts\")\n",
    "\n",
    "sc.pl.umap(adata_combat, color=combat_marker_genes['DAM'])\n",
    "\n",
    "sc.pl.umap(adata_combat, color=combat_marker_genes['Homeo'])\n",
    "\n",
    "sc.pl.umap(adata_combat, color=combat_marker_genes['Macroph'])\n",
    "\n",
    "sc.pl.umap(adata_combat, color= [\"genotype\", \"treatment\"], wspace= 0.5, legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0.01\n",
    "while i <= 2.1:\n",
    "    sc.tl.leiden(\n",
    "        adata,\n",
    "        resolution=i,\n",
    "        random_state=0,\n",
    "        n_iterations=2,\n",
    "        directed=False,\n",
    "        key_added=f\"leiden_{i}\",\n",
    "        flavor=\"igraph\"\n",
    "    )\n",
    "    i += 0.01\n",
    "    i = round(i, 2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=[\"leiden_0.07\", \"leiden_0.3\", \"leiden_0.7\", \"leiden_1.0\", \"leiden_1.3\",\"leiden_1.6\",\"leiden_1.8\",\"leiden_2.1\",], legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color='leiden_1.6', legend_loc=\"on data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Annotation\n",
    "### Automated Annotation\n",
    "#### Decoupler & PanglaoDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = dc.get_resource('PanglaoDB')\n",
    "# Filter by canonical_marker and mouse\n",
    "markers['mouse_sensitivity'] = markers['mouse_sensitivity'].astype(float)\n",
    "markers = markers[(markers['mouse'] == 'True') & markers['organ'].isin(['Brain', 'Immune system', 'Blood']) & (markers['mouse_sensitivity'] > 0.5)]\n",
    "# Remove duplicated entries\n",
    "markers = markers[~markers.duplicated(['cell_type', 'genesymbol'])]\n",
    "# sort highest to lowest 'mouse sensitivity'\n",
    "markers = markers.sort_values('mouse_sensitivity', ascending=False)\n",
    "markers['genesymbol'] = markers['genesymbol'].str.capitalize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dc.run_ora(\n",
    "    mat=adata,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=1,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acts = dc.get_acts(adata, obsm_key='ora_estimate')\n",
    "\n",
    "# We need to remove inf and set them to the maximum value observed for pvals=0\n",
    "acts_v = acts.X.ravel()\n",
    "max_e = np.nanmax(acts_v[np.isfinite(acts_v)])\n",
    "acts.X[~np.isfinite(acts.X)] = max_e\n",
    "acts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.pl.umap(acts, color=['Microglia', 'Macrophages', 'Platelets', 'Dendritic cells', 'Monocytes','Neural stem/precursor cells'], cmap='RdBu_r', legend_loc=\"on data\", vmax=25)\n",
    "\n",
    "sc.pl.violin(acts, keys=['Microglia','Macrophages'], groupby='leiden_0.7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = dc.rank_sources_groups(acts, groupby='leiden_0.7', reference='rest', method='t-test_overestim_var')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_ctypes = 3\n",
    "ctypes_dict = df.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "ctypes_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_clusters = {\n",
    "    '0': 'Homeostatic Microglia I',\n",
    "    '10': 'Low-ribosomal Microglia',\n",
    "    '2': 'Homeostatic Microglia I',\n",
    "    '3': 'Homeostatic Microglia I',\n",
    "    '4': 'Homeostatic Microglia I',\n",
    "    '5': 'Homeostatic Microglia I',\n",
    "    '17': 'Homeostatic Microglia I',\n",
    "    '6': 'High ribosomal DAM',\n",
    "    '7': 'High ribosomal DAM',\n",
    "    '8': 'High ribosomal DAM',\n",
    "    '9': 'Low-ribosomal Microglia',\n",
    "    '11': 'Homeostatic Microglia II',\n",
    "    '21': 'High ribosomal DAM',\n",
    "    '12': 'DAM',\n",
    "    '13': 'DAM',\n",
    "    '14': 'DAM',\n",
    "    '15': 'Macrophage',\n",
    "    '16': 'Homeostatic Microglia I',\n",
    "    '18': 'Macrophage',\n",
    "    '1': 'Homeostatic Microglia II',\n",
    "    '19': 'High ribosomal DAM',\n",
    "    '20': 'High ribosomal DAM',\n",
    "}\n",
    "\n",
    "adata.obs['Man_annotation'] = (\n",
    "adata.obs['leiden_1.6']\n",
    ".map(merge_clusters)\n",
    ".astype('category')\n",
    ")\n",
    "\n",
    "Man_anno_counts = adata.obs['Man_annotation'].value_counts()\n",
    "print(Man_anno_counts)\n",
    "unique_values = set(merge_clusters.values())\n",
    "print(unique_values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, groupby=\"Man_annotation\", method='wilcoxon')\n",
    "# Get the unique groups\n",
    "groups = adata.uns['rank_genes_groups']['names'].dtype.names\n",
    "\n",
    "# Iterate over the groups\n",
    "for group in groups:\n",
    "    # Get the top 5 genes for the current group\n",
    "    df = sc.get.rank_genes_groups_df(adata, group=group).head(10)\n",
    "    # Join the gene names with commas\n",
    "    genes = ', '.join(df['names'])\n",
    "    # Print the group name and the gene names\n",
    "    print(f\"Group_{group}:{genes}\")\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "\n",
    "# Assuming 'adata' is already loaded and prepared with 'Man_annotation' as a categorical variable\n",
    "\n",
    "# Define the color palette\n",
    "palette = ['#8FBC8F', '#F08080', '#B0E0E6', '#DDA0DD', '#BC8F8F', '#DAA520']\n",
    "\n",
    "# Generate the UMAP plot with the defined palette\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "sc.pl.umap(adata, color='Man_annotation', size=15, ax=ax, show=False, title='', palette=palette, legend_loc='on data', legend_fontsize=6, colorbar_loc='right')\n",
    "\n",
    "# Set the plot title and axes labels\n",
    "ax.set_xlabel('UMAP 1', fontsize=16, labelpad=5)\n",
    "ax.set_ylabel('UMAP 2', fontsize=16, labelpad=5)\n",
    "\n",
    "ax.set_aspect('equal', adjustable='datalim')\n",
    "\n",
    "# Retrieve the categories from 'adata'\n",
    "categories = adata.obs['Man_annotation'].cat.categories\n",
    "\n",
    "# Create a dictionary of categories and their corresponding colors from the palette\n",
    "color_dict = dict(zip(categories, palette))\n",
    "\n",
    "# Legend descriptions\n",
    "legend_descriptions = {\n",
    "    'DAM': 'Number of cells: 2596 \\nTop DEGs: Ctsz, Cd9, Ctsd, Trem2, Ctsb, Ctsl',\n",
    "    'High ribosomal DAM': 'Number of cells: 4725 \\nTop DEGs: Fth1, Eef1a1, Ftl1, Rpl21, Rpl41, Fau',\n",
    "    'Homeostatic Microglia I': 'Number of cells: 6018 \\nTop DEGs: Cx3cr1, P2ry12, Selplg, Tmem119, Lgmn, Serinc3',\n",
    "    'Homeostatic Microglia II': 'Number of cells: 1499 \\nTop DEGs: Cst3, Rplp1, Selenop, mt-Co3, Rps29, mt-Cytb',\n",
    "    'Low-ribosomal Microglia': 'Number of cells: 1744 \\nTop DEGs: Xist, Malat1, Son, Srgap2, Tgfbr1, Fus',\n",
    "    'Macrophage': 'Number of cells: 780 \\nTop DEGs: Pf4, Mrc1, F13a1, Ms4a7, Dab2, Lyz2'\n",
    "}\n",
    "\n",
    "# Create patches for each category for the detailed legend\n",
    "patches = [mpatches.Patch(color=color_dict[cat], label=f\"{cat}\\n{legend_descriptions[cat]}\") for cat in categories]\n",
    "\n",
    "# Display the detailed legend to the right of the main figure\n",
    "legend = ax.legend(handles=patches, loc='center left', bbox_to_anchor=(1, 0.5), frameon=True, fontsize=10, title=\"Cluster Characteristics\", handleheight=1.2, handlelength=1.5)\n",
    "plt.setp(legend.get_title(), fontsize='12')\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 0.8, 1])  # Adjust layout to make room for legend\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting and Visualizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(10, 5))  # Adjust figsize as needed\n",
    "\n",
    "# Plot the UMAP for \"sample\"\n",
    "sc.pl.umap(adata, color=\"sample\", size=12, ax=axs[0], show=False)\n",
    "axs[0].set_title(\"Cell Visualization by Dataset\", fontsize=16)\n",
    "\n",
    "# Plot the UMAP for \"genotype\"\n",
    "sc.pl.umap(adata, color=\"genotype\", size=18, ax=axs[1], show=False)\n",
    "axs[1].set_title(\"Cell Visualization by Genotype\", fontsize=16)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of genes for the first and second UMAP plots\n",
    "genes_1 = [\"Apoe\", \"Trem2\", \"Lpl\", \"Tyrobp\"]\n",
    "genes_2 = [\"Cx3cr1\", \"P2ry12\", \"Olfml3\", \"Tmem119\"]\n",
    "genes_3 = [\"Mrc1\", \"Lyz2\", \"Cd68\",\"Ccr2\"]\n",
    "\n",
    "# Create a figure with two rows and four columns of subplots\n",
    "fig, axs = plt.subplots(3, 4, figsize=(24, 16))  # Adjust figsize as needed\n",
    "\n",
    "# Plot the first set of genes\n",
    "for i, gene in enumerate(genes_1):\n",
    "    sc.pl.umap(adata, color=gene, vmax=4, size=30, ax=axs[0, i], show=False)\n",
    "    axs[0, i].set_title(f\"{gene}\", fontsize=26)\n",
    "\n",
    "# Plot the second set of genes\n",
    "for i, gene in enumerate(genes_2):\n",
    "    sc.pl.umap(adata, color=gene, vmax=4, size=30, ax=axs[1, i], show=False)\n",
    "    axs[1, i].set_title(f\"{gene}\", fontsize=26)\n",
    "\n",
    "for i, gene in enumerate(genes_3):\n",
    "    sc.pl.umap(adata, color=gene, vmax=4, size=30, ax=axs[2, i], show=False)\n",
    "    axs[2, i].set_title(f\"{gene}\", fontsize=26)\n",
    "\n",
    "\n",
    "fig.text(0.5, 0.95, 'DAM-associated Genes', ha='center', fontsize=28)\n",
    "fig.text(0.5, 0.63, 'Homeostatic-associated Genes', ha='center', fontsize=28)\n",
    "fig.text(0.5, 0.31, 'Canonical Macrophage Marker Genes', ha='center', fontsize=28)\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "plt.subplots_adjust(hspace=0.3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata, marker_genes['Nfkb'], groupby='leiden_1.4', save='_Nfkb.png', show=False)\n",
    "\n",
    "# Create and save the second heatmap\n",
    "sc.pl.heatmap(adata, marker_genes['Mapk'], groupby='leiden_1.4', save='_Mapk.png', show=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(adata, ribo_genes, groupby='leiden_1.4', show=False)\n",
    "fig1 = plt.gcf()\n",
    "ax1 = plt.gca()\n",
    "\n",
    "# Add title and axis labels to the first heatmap\n",
    "ax1.set_xlabel('Ribosomal Genes', fontsize=14, labelpad=20)\n",
    "ax1.set_ylabel('Cells Grouped in Clusters', fontsize=14, labelpad=10)\n",
    "ax1.xaxis.set_label_coords(20, -0.05)\n",
    "\n",
    "# Enhance x and y ticks for better readability\n",
    "plt.xticks(rotation=90, fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "\n",
    "# Optimize layout\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.95])\n",
    "\n",
    "# Save the first heatmap as an image\n",
    "fig1.savefig('figures/heatmap_ribo.png', bbox_inches='tight')\n",
    "\n",
    "sc.pl.heatmap(adata, marker_genes['Nfkb'], groupby='leiden_1.4', save='_Nfkb.png', show=False)\n",
    "\n",
    "# Create and save the second heatmap\n",
    "sc.pl.heatmap(adata, marker_genes['Mapk'], groupby='leiden_1.4', save='_Mapk.png', show=False)\n",
    "\n",
    "# Load the saved images\n",
    "img1 = Image.open('figures/heatmap_ribo.png')\n",
    "img2 = Image.open('figures/heatmap_Nfkb.png')\n",
    "img3 = Image.open('figures/heatmap_Mapk.png')\n",
    "\n",
    "# Create a figure with a GridSpec layout\n",
    "fig = plt.figure(figsize=(16, 16))\n",
    "gs = gridspec.GridSpec(2, 2, height_ratios=[1, 1])\n",
    "\n",
    "# Display the ribosomal genes heatmap in the centered position of the first row\n",
    "ax0 = fig.add_subplot(gs[0, :])\n",
    "ax0.imshow(img1)\n",
    "ax0.axis('off')\n",
    "ax0.set_title('Heatmap of Ribosomal Genes Grouped by Leiden Clusters', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Display the NFKB genes heatmap in the first subplot of the second row\n",
    "ax1 = fig.add_subplot(gs[1, 0])\n",
    "ax1.imshow(img2)\n",
    "ax1.axis('off')\n",
    "ax1.set_title('Heatmap of NFKB Genes Grouped by Leiden Clusters', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Display the MAPK genes heatmap in the second subplot of the second row\n",
    "ax2 = fig.add_subplot(gs[1, 1])\n",
    "ax2.imshow(img3)\n",
    "ax2.axis('off')\n",
    "ax2.set_title('Heatmap of MAPK Genes Grouped by Leiden Clusters', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Adjust layout to ensure titles and bottoms are aligned\n",
    "fig.align_labels()  # Align labels\n",
    "fig.align_xlabels([ax0, ax1, ax2])  # Align x-axis labels\n",
    "\n",
    "# Display the plots\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata.obs[['phase', 'Man_annotation']]\n",
    "\n",
    "prop_df = df.groupby(['Man_annotation', 'phase']).size().unstack(fill_value=0)\n",
    "prop_df = prop_df.div(prop_df.sum(axis=1), axis=0)\n",
    "\n",
    "# Plotting with improved visual appeal\n",
    "fig, ax = plt.subplots(figsize=(12, 6))\n",
    "prop_df.plot(kind='bar', stacked=True, ax=ax, colormap='tab20')\n",
    "\n",
    "# Removing the grid\n",
    "ax.grid(False)\n",
    "\n",
    "# Additional improvements\n",
    "ax.set_ylabel('Proportion', fontsize=14)\n",
    "ax.set_title('Distribution of Cell Cycle Phases Across Clusters', fontsize=16, fontweight='bold')\n",
    "ax.legend(title='Cell Cycle Phase', bbox_to_anchor=(1.05, 1), loc='upper left', fontsize=12)\n",
    "ax.set_xlabel('Cluster', fontsize=14)\n",
    "plt.xticks(rotation=45, ha='right', fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "\n",
    "# Enhancing the layout\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
